File: src/Scanner.Core/Checks/FixedWidthLayoutCheck.cs
using System.Globalization;
using System.Text.RegularExpressions;
using Scanner.Core.Rules;

namespace Scanner.Core.Checks;

/// <summary>
/// Checks for fixed-width layouts that can block reflow.
/// </summary>
public sealed class FixedWidthLayoutCheck : ICheck
{
    private static readonly Regex TagRegex = new("<(?<tag>[a-zA-Z0-9:-]+)(?<attrs>[^>]*)>", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    /// <summary>
    /// Gets the unique identifier for this check.
    /// </summary>
    public static string Id => "fixed-width-layout";

    /// <inheritdoc />
    string ICheck.Id => Id;

    /// <inheritdoc />
    public IReadOnlyCollection<string> ApplicableKinds { get; } = new[] { "html", "htm", "cshtml", "razor", "xaml" };

    /// <inheritdoc />
    public IEnumerable<Issue> Run(CheckContext context, RuleDefinition rule)
    {
        if (context.Kind.Equals("xaml", StringComparison.OrdinalIgnoreCase))
        {
            return RunXaml(context, rule);
        }

        return RunMarkup(context, rule);
    }

    private static readonly string[] StyleWidthProperties = { "width", "min-width" };

    private static IEnumerable<Issue> RunMarkup(CheckContext context, RuleDefinition rule)
    {
        foreach (Match match in TagRegex.Matches(context.Content))
        {
            var attrs = match.Groups["attrs"].Value;
            var style = AttributeParser.GetAttributeValue(attrs, "style");
            if (TryGetFixedStyleWidth(style, out var propertyName, out var propertyValue))
            {
                var styleLineNumber = TextUtilities.GetLineNumber(context.Content, match.Index);
                var description = $"Element uses fixed {DescribeProperty(propertyName)} ({propertyName}: {propertyValue}).";
                yield return new Issue(rule.Id, Id, context.FilePath, styleLineNumber, description, match.Value);
                continue;
            }

            var widthAttribute = AttributeParser.GetAttributeValue(attrs, "width");
            if (!IsFixedMarkupLength(widthAttribute))
            {
                continue;
            }

            var widthLineNumber = TextUtilities.GetLineNumber(context.Content, match.Index);
            yield return new Issue(
                rule.Id,
                Id,
                context.FilePath,
                widthLineNumber,
                $"Element uses a fixed width attribute (width=\"{widthAttribute}\").",
                match.Value);
        }
    }

    private static IEnumerable<Issue> RunXaml(CheckContext context, RuleDefinition rule)
    {
        foreach (Match match in TagRegex.Matches(context.Content))
        {
            var attrs = match.Groups["attrs"].Value;
            var widthValue = AttributeParser.GetAttributeValue(attrs, "Width");
            if (IsFixedWidthValue(widthValue))
            {
                var widthLineNumber = TextUtilities.GetLineNumber(context.Content, match.Index);
                yield return new Issue(
                    rule.Id,
                    Id,
                    context.FilePath,
                    widthLineNumber,
                    $"XAML element uses fixed width (Width=\"{widthValue}\").",
                    match.Value);
                continue;
            }

            var minWidthValue = AttributeParser.GetAttributeValue(attrs, "MinWidth");
            if (!IsFixedWidthValue(minWidthValue))
            {
                continue;
            }

            var minWidthLineNumber = TextUtilities.GetLineNumber(context.Content, match.Index);
            yield return new Issue(
                rule.Id,
                Id,
                context.FilePath,
                minWidthLineNumber,
                $"XAML element uses fixed minimum width (MinWidth=\"{minWidthValue}\").",
                match.Value);
        }
    }

    private static bool IsFixedWidthValue(string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return false;
        }

        if (value.Equals("Auto", StringComparison.OrdinalIgnoreCase))
        {
            return false;
        }

        if (value.Contains('*'))
        {
            return false;
        }

        return double.TryParse(value, NumberStyles.Float, CultureInfo.InvariantCulture, out var number) && number > 0;
    }

    private static bool TryGetFixedStyleWidth(string? style, out string propertyName, out string propertyValue)
    {
        foreach (var property in StyleWidthProperties)
        {
            var value = StyleUtilities.GetLastPropertyValue(style, property);
            if (!string.IsNullOrWhiteSpace(value) && StyleUtilities.IsFixedLength(value))
            {
                propertyName = property;
                propertyValue = value;
                return true;
            }
        }

        propertyName = string.Empty;
        propertyValue = string.Empty;
        return false;
    }

    private static bool IsFixedMarkupLength(string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return false;
        }

        var trimmed = value.Trim();
        if (trimmed.EndsWith('%'))
        {
            return false;
        }

        if (StyleUtilities.IsFixedLength(trimmed))
        {
            return true;
        }

        return double.TryParse(trimmed, NumberStyles.Float, CultureInfo.InvariantCulture, out var number) && number > 0;
    }

    private static string DescribeProperty(string propertyName)
    {
        return propertyName.Equals("min-width", StringComparison.OrdinalIgnoreCase)
            ? "minimum width"
            : "width";
    }
}

File: node_engine/src/checks/FixedWidthLayoutCheck.js
const { getAttributeValue } = require("./AttributeParser");
const { getLineNumber } = require("./TextUtilities");
const { getLastPropertyValue, isFixedLength } = require("./StyleUtilities");

const tagRegex = /<(?<tag>[a-zA-Z0-9:-]+)(?<attrs>[^>]*)>/gi;
const styleWidthProperties = ["width", "min-width"];
const styleHeightProperties = ["height", "min-height"];
const MIN_VIEWPORT_WIDTH_PX = 320;
const MIN_VIEWPORT_HEIGHT_PX = 256;
const BASE_FONT_SIZE_PX = 16;

const lengthRegex = /^(?<value>-?\d+(?:\.\d+)?)(?<unit>px|pt|pc|cm|mm|in|em|rem)?$/i;

const isFixedDimensionValue = (value, minPx) => {
  if (!value || !value.trim()) {
    return false;
  }

  if (value.toLowerCase() === "auto") {
    return false;
  }

  if (value.includes("*")) {
    return false;
  }

  const parsed = Number.parseFloat(value);
  return !Number.isNaN(parsed) && parsed > minPx;
};

const isFixedWidthValue = (value) => isFixedDimensionValue(value, MIN_VIEWPORT_WIDTH_PX);

const convertToPx = (number, unit) => {
  switch (unit) {
    case "px":
      return number;
    case "pt":
      return (number * 96) / 72;
    case "pc":
      return number * 16;
    case "in":
      return number * 96;
    case "cm":
      return (number * 96) / 2.54;
    case "mm":
      return (number * 96) / 25.4;
    case "em":
    case "rem":
      return number * BASE_FONT_SIZE_PX;
    default:
      return null;
  }
};

const parseFixedLengthToPx = (value, { assumePixels = false } = {}) => {
  if (!value || !value.trim()) {
    return null;
  }

  const trimmed = value.trim();
  if (trimmed.endsWith("%")) {
    return null;
  }

  if (isFixedLength(trimmed)) {
    const match = lengthRegex.exec(trimmed);
    if (!match) {
      return null;
    }

    const number = Number.parseFloat(match.groups.value);
    if (Number.isNaN(number) || number <= 0) {
      return null;
    }

    const unit = match.groups.unit ? match.groups.unit.toLowerCase() : null;
    const unitToUse = unit ?? (assumePixels ? "px" : null);
    if (!unitToUse) {
      return null;
    }

    return convertToPx(number, unitToUse);
  }

  const parsed = Number.parseFloat(trimmed);
  if (Number.isNaN(parsed) || parsed <= 0) {
    return null;
  }

  if (assumePixels) {
    return parsed;
  }

  return null;
};

const isFixedLengthAtLeast = (value, minPx, options) => {
  const size = parseFixedLengthToPx(value, options);
  if (size == null) {
    return false;
  }

  return size > minPx;
};

const tryGetFixedStyleDimension = (style, properties, minPx) => {
  for (const property of properties) {
    const value = getLastPropertyValue(style, property);
    if (!value) {
      continue;
    }

    const size = parseFixedLengthToPx(value);
    if (size != null && size > minPx) {
      return { propertyName: property, propertyValue: value, pixelValue: size };
    }
  }

  return null;
};

const tryGetFixedStyleWidth = (style) => tryGetFixedStyleDimension(style, styleWidthProperties, MIN_VIEWPORT_WIDTH_PX);
const tryGetFixedStyleHeight = (style) => tryGetFixedStyleDimension(style, styleHeightProperties, MIN_VIEWPORT_HEIGHT_PX);

const isFixedMarkupLength = (value, { minPx = 0, assumePixels = true } = {}) =>
  isFixedLengthAtLeast(value, minPx, { assumePixels });

const describeProperty = (propertyName) => {
  const normalized = propertyName.toLowerCase();
  if (normalized === "min-width") {
    return "minimum width";
  }
  if (normalized === "min-height") {
    return "minimum height";
  }
  return normalized.includes("height") ? "height" : "width";
};

const runMarkup = (context, rule) => {
  const issues = [];
  for (const match of context.content.matchAll(tagRegex)) {
    const attrs = match.groups.attrs;
    const style = getAttributeValue(attrs, "style");
    const fixedStyleWidth = tryGetFixedStyleWidth(style);
    if (fixedStyleWidth) {
      const line = getLineNumber(context.content, match.index);
      issues.push({
        ruleId: rule.id,
        checkId: FixedWidthLayoutCheck.id,
        filePath: context.filePath,
        line,
        message: `Element uses fixed ${describeProperty(fixedStyleWidth.propertyName)} (${fixedStyleWidth.propertyName}: ${fixedStyleWidth.propertyValue}).`,
        evidence: match[0]
      });
      continue;
    }

    const fixedStyleHeight = tryGetFixedStyleHeight(style);
    if (fixedStyleHeight) {
      const line = getLineNumber(context.content, match.index);
      issues.push({
        ruleId: rule.id,
        checkId: FixedWidthLayoutCheck.id,
        filePath: context.filePath,
        line,
        message: `Element uses fixed ${describeProperty(fixedStyleHeight.propertyName)} (${fixedStyleHeight.propertyName}: ${fixedStyleHeight.propertyValue}).`,
        evidence: match[0]
      });
      continue;
    }

    const widthAttribute = getAttributeValue(attrs, "width");
    if (isFixedMarkupLength(widthAttribute, { minPx: MIN_VIEWPORT_WIDTH_PX })) {
      const line = getLineNumber(context.content, match.index);
      issues.push({
        ruleId: rule.id,
        checkId: FixedWidthLayoutCheck.id,
        filePath: context.filePath,
        line,
        message: `Element uses a fixed width attribute (width="${widthAttribute}").`,
        evidence: match[0]
      });
      continue;
    }

    const heightAttribute = getAttributeValue(attrs, "height");
    if (isFixedMarkupLength(heightAttribute, { minPx: MIN_VIEWPORT_HEIGHT_PX })) {
      const line = getLineNumber(context.content, match.index);
      issues.push({
        ruleId: rule.id,
        checkId: FixedWidthLayoutCheck.id,
        filePath: context.filePath,
        line,
        message: `Element uses a fixed height attribute (height="${heightAttribute}").`,
        evidence: match[0]
      });
    }
  }

  return issues;
};

const runXaml = (context, rule) => {
  const issues = [];
  for (const match of context.content.matchAll(tagRegex)) {
    const attrs = match.groups.attrs;
    const widthValue = getAttributeValue(attrs, "Width");
    if (isFixedDimensionValue(widthValue, MIN_VIEWPORT_WIDTH_PX)) {
      const line = getLineNumber(context.content, match.index);
      issues.push({
        ruleId: rule.id,
        checkId: FixedWidthLayoutCheck.id,
        filePath: context.filePath,
        line,
        message: `XAML element uses fixed width (Width="${widthValue}").`,
        evidence: match[0]
      });
      continue;
    }

    const minWidthValue = getAttributeValue(attrs, "MinWidth");
    if (isFixedDimensionValue(minWidthValue, MIN_VIEWPORT_WIDTH_PX)) {
      const line = getLineNumber(context.content, match.index);
      issues.push({
        ruleId: rule.id,
        checkId: FixedWidthLayoutCheck.id,
        filePath: context.filePath,
        line,
        message: `XAML element uses fixed minimum width (MinWidth="${minWidthValue}").`,
        evidence: match[0]
      });
      continue;
    }

    const heightValue = getAttributeValue(attrs, "Height");
    if (isFixedDimensionValue(heightValue, MIN_VIEWPORT_HEIGHT_PX)) {
      const line = getLineNumber(context.content, match.index);
      issues.push({
        ruleId: rule.id,
        checkId: FixedWidthLayoutCheck.id,
        filePath: context.filePath,
        line,
        message: `XAML element uses fixed height (Height="${heightValue}").`,
        evidence: match[0]
      });
      continue;
    }

    const minHeightValue = getAttributeValue(attrs, "MinHeight");
    if (!isFixedDimensionValue(minHeightValue, MIN_VIEWPORT_HEIGHT_PX)) {
      continue;
    }

    const line = getLineNumber(context.content, match.index);
    issues.push({
      ruleId: rule.id,
      checkId: FixedWidthLayoutCheck.id,
      filePath: context.filePath,
      line,
      message: `XAML element uses fixed minimum height (MinHeight="${minHeightValue}").`,
      evidence: match[0]
    });
  }

  return issues;
};

const FixedWidthLayoutCheck = {
  id: "fixed-width-layout",
  applicableKinds: ["html", "htm", "cshtml", "razor", "xaml"],
  run(context, rule) {
    if (context.kind.toLowerCase() === "xaml") {
      return runXaml(context, rule);
    }

    return runMarkup(context, rule);
  }
};

module.exports = {
  FixedWidthLayoutCheck,
  isFixedDimensionValue,
  isFixedWidthValue,
  tryGetFixedStyleWidth,
  tryGetFixedStyleHeight,
  isFixedMarkupLength,
  parseFixedLengthToPx,
  isFixedLengthAtLeast,
  describeProperty,
  MIN_VIEWPORT_WIDTH_PX,
  MIN_VIEWPORT_HEIGHT_PX
};
